crates/miden-lib/build.rs

```diff
@@ -779,90 +779,131 @@ impl TxKernelErrorCategory {
+// EVENT CONSTANTS FILE GENERATION
+// ================================================================================================
+
+/// Reads all MASM files from the `asm_source_dir` and extracts event definitions,
+/// then generates the transaction_events.rs file with constants.
+fn generate_event_constants(asm_source_dir: &Path, target_dir: &Path) -> Result<()> {
+    // Extract all event definitions from MASM files
+    let events = extract_all_event_definitions(asm_source_dir)?;
+
+    // Generate the events file in OUT_DIR
+    let event_file_content = generate_event_file_content(&events)?;
+    let event_file_path = target_dir.join("transaction_events.rs");
+    fs::write(event_file_path, event_file_content).into_diagnostic()?;
+
+    Ok(())
+}
+
+/// Extract all `const.X=event("x")` definitions from all MASM files
+fn extract_all_event_definitions(asm_source_dir: &Path) -> Result<BTreeMap<String, String>> {
+    // collect mappings event path to const variable name, we want a unique mapping
+    // which we use to generate the constants and enum variant names
+    let mut events = BTreeMap::new();
+
+    // Walk all MASM files
+    for entry in WalkDir::new(asm_source_dir) {
+        let entry = entry.into_diagnostic()?;
+        if !is_masm_file(entry.path()).into_diagnostic()? {
+            continue;
+        }
+        let file_contents = fs::read_to_string(entry.path()).into_diagnostic()?;
+        extract_event_definitions_from_file(&mut events, &file_contents, entry.path())?;
     }
 
-    #[cfg(feature = "std")]
-    {
-        writeln!(&mut f, "// This file is generated by build.rs, do not modify")?;
+    Ok(events)
+}
+
+/// Extract `const.${X}=event("${x::path}")` definitions from a single MASM file
+fn extract_event_definitions_from_file(
+    events: &mut BTreeMap<String, String>,
+    file_contents: &str,
+    file_path: &Path,
+) -> Result<()> {
+    let regex = Regex::new(r#"const\.(\w+)=event\("([^"]+)"\)"#).unwrap();
+
+    for capture in regex.captures_iter(file_contents) {
+        let Some(const_name) = capture.get(1) else { continue };
+        let Some(event_path) = capture.get(2) else { continue };
+
+        let event_path = event_path.as_str();
+        let const_name = const_name.as_str();
+
+        let const_name_wo_suffix =
+            if let Some((const_name_wo_suffix, _)) = const_name.rsplit_once("_EVENT") {
+                const_name_wo_suffix.to_string()
+            } else {
+                const_name.to_owned()
+            };
 
-        // we require a form of lut, but we don't want to bother supporting non-std cases for now
-        let mut inner = String::new();
-        for (name, (_, value)) in map.iter() {
-            inner.push_str(format!("                   ({value}, \"{name}\"),\n").as_str());
+        if let Some((first, _)) = event_path.split_once("::") {
+            if first != "miden" {
+                // i.e. stdlib we don't want to have here
+            }
```

Comment:
Related to one of the previous comments: if we do come across a `stdlib` event, would be awesome to check if `miden-stdlib` actually exposes it. Though, I'm not sure if this is currently possible.

Regardless, I think we should make sure that the events either start with `miden` or `stdlib` (if we come across something else, that's probably an error).